pyrealm/demography/canopy.py

```diff
@@ -0,0 +1,133 @@
+"""Functionality for canopy modelling."""
+
+import numpy as np
+from numpy.typing import NDArray
+from scipy.optimize import root_scalar
+
+from pyrealm.demography.canopy_functions import (
+    calculate_relative_canopy_radius_at_z,
+    calculate_stem_projected_crown_area_at_z,
+    calculate_stem_projected_leaf_area_at_z,
+    solve_community_projected_canopy_area,
+)
+from pyrealm.demography.community import Community
+
+
+class Canopy:
+    """A class containing attributes of a canopy, including the structure.
+
+    at which canopy layers close. This also includes
+        ground level, which will capture all remaning crown and leaf area below the last
+        closed layer.
+    """
+
+    def __init__(
+        self,
+        community: Community,
+        canopy_gap_fraction: float,
+        layer_tolerance: float = 0.001,
+    ) -> None:
+        # Calculate community wide properties: total crown area and maximum height
+        self.total_community_crown_area: float = (
+            community.cohort_data["crown_area"] * community.cohort_data["n_individuals"]
+        ).sum()
+        """Total crown area across individuals in the community (metres 2)."""
+
+        self.max_stem_height: float = community.cohort_data["stem_height"].max()
+        """Maximum height of any individual in the community (metres)."""
+
+        self.crown_area_per_layer: float = community.cell_area * (
+            1 - canopy_gap_fraction
+        )
+        """Total crown area permitted in a single canopy layer, given the available
+        cell area of the community and its canopy gap fraction."""
+
+        self.n_layers: int = int(
+            np.ceil(self.total_community_crown_area / self.crown_area_per_layer)
+        )
+        """Total number of canopy layers."""
+
+        self.n_cohorts: int = community.number_of_cohorts
+        """Total number of cohorts in the canopy."""
+
+        # Find the closure heights of the canopy layers under the perfect plasticity
+        # approximation by solving Ac(z) - L_n = 0 across the community where L is the
+        # total cumulative crown area in layer n and above, discounted by the canopy gap
+        # fraction.
+
+        self.layer_heights: NDArray[np.float32] = np.zeros(
+            (self.n_layers, 1), dtype=np.float32
+        )
+        """Column vector of the heights of canopy layers."""
+
+        # Loop over the layers except for the final layer, which will be the partial
+        # remaining vegetation below the last closed layer.
+        starting_guess = self.max_stem_height
+        for layer in np.arange(self.n_layers - 1):
+            target_area = (layer + 1) * community.cell_area * (1 - canopy_gap_fraction)
+
+            # TODO - the solution here is predictably closer to the upper bracket, might
```

Comment:
Instead of putting a TODO in the code, creating an issue has a higher likelihood of this being seen and solved.